### ISSUE TITLE
LC0068 - False Negative RecordRef

### ISSUE THREAD

**Issue by pri-kise** [StefanMaron/BusinessCentral.LinterCop#755]
I know that it's difficult to check this rule for RecordRef, but there is one case, that maybe can be found.

```al
local procedure DoMagic(Entry: Record "PTE Entry"; )
var
    IncomingDocument: Record "Incoming Document";
    RecRef: RecordRef;
begin
    if Entry."Record ID".TableNo() <> Database::"Incoming Document" then
        exit;

    RecRef.Open(Database::"Incoming Document"); 
    if CompanyName() <> Entry."Company Name" then begin
        RecRef.ChangeCompany(Entry."Company Name");
        IncomingDocument.ChangeCompany(Entry."Company Name");
    end;
    if not RecRef.Get(Entry."Record ID") then //missing warning
        exit;
    RecRef.SetTable(IncomingDocument);
    //more Magic
end;
```
At the least the indirect read permission could be found here.

**Comment by StefanMaron** [comment]
I dont see any option to see what table the reocrdref is pointing at, maybe @Arthurvdv has an idea?

**Comment by Arthurvdv** [comment]
This is a tough one, not impossible I believe, but could be complex.

The assignment of the table on the RecordRef is determined during runtime. So the LinterCop, as a static code analyzer, can’t determine this during compile time.

What I can think of is something similar as the CodeCop [AA0181](https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/analyzers/codecop-aa0181) warning. This would then translate to;

![image](https://github.com/user-attachments/assets/9f7e7059-a4d6-4dff-a186-b51cbb23d4e7)

1)	Trigger on a .Get() of a RecordRef instance
2)	Retrieving the Local/Global variable instance
3)	Loop through the codeblock of the method, line-by-line, for a match on .Open()
4)	Retrieve the Database:: value

Then there could be more complex situations:

```AL
local procedure DoMagic(Entry: Record "PTE Entry")
var
    RecRef: RecordRef;
begin
    RecRef.Open(Database::"Incoming Document");
    RecRef.Get(Entry."Record ID");

    RecRef.Open(Database::"Document Sharing");
    RecRef.Get(Entry."Record ID");
    RecRef.Modify(false)
end;
```

The .Modify() here could then raise a false positive for the "Incoming Document" table. So when looping through the codeblock, we need to start at the line of the action (.Get/.Modify) and walk upwards till we find the first matching .Open() reference.

With the restriction that the .Open operator and .Get/.Modift/.. operator are within in the same procedure I think it should be possible. I'm unsure if there are other scenario's where the "just-look-at-the-lines-above-in-the-code" method could still cause false positives.

**Comment by Arthurvdv** [comment]
Hi, over the past months I’ve been putting a lot of time and care into the successor of BusinessCentral.LinterCop: ALCops (https://alcops.dev/). Because of that, this repository is no longer actively maintained.

I’m closing this issue, as it won’t be handled here anymore. If this is still relevant for you, please have a look at ALCops/Analyzers (https://github.com/ALCops/Analyzers).

To make the transition easier, you can also follow the LinterCop Migration Guide: https://alcops.dev/docs/lintercop-migration/

If you have any questions or run into issues during the transition, feel free to reach out through ALCops. Happy to help 🙂
